### PR TITLE
Make NodeTaskRequest tasks cancellable

### PR DESCRIPTION
**Pull request by benwtrent**
We pass the actionTask to the subsequent task action so that the task action can be cancelled if the upstream tasks are cancelled.

But, the actionTask passed is created by the `NodeTaskRequest`, which isn't cancellable. 

This commit addresses this discrepancy and allows for down stream task actions to be cancelled if the upstream http connections are closed.

Related to: #88081

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java [hunk 1]
```diff
@@ -323,6 +328,9 @@ class NodeTransportHandler implements TransportRequestHandler<NodeTaskRequest> {
 
         @Override
         public void messageReceived(final NodeTaskRequest request, final TransportChannel channel, Task task) throws Exception {
+            if (task instanceof CancellableTask cancellableTask) {
+                cancellableTask.ensureNotCancelled();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:332]: Is this needed? The transport layer will have just checked for cancellation before calling `messageReceived` I think.

**Review by DaveCTurner — COMMENTED** [review]
I left one comment inline. Also could we have a test for this?

File: server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:332]: @DaveCTurner I am unfamiliar where this code is called within the transport layer. 

I copied this logic loosely based on NodeTransportHandler.

https://github.com/elastic/elasticsearch/blob/2f5ec518f349db7e446c0d4696666284bd983357/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java#L327-L335

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:332]: I will happily remove it. The big change I need for subsequent changes is that the task is cancellable, I thought it prudent to check here. But if its done for sure higher up the stack, then I am good with removing it

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:332]: > I copied this logic loosely based on NodeTransportHandler.

`git blame` --> guilty as charged 🤡 

Yeah this is not necessary, we're called from `RequestHandlerRegistry#processMessageReceived` which just a few lines earlier obtained its task from `TaskManager#register`, so called `TaskManager#registerCancellableTask`, and this method would have thrown if the task were already cancelled.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:332]: @DaveCTurner I will remove it then :D thanks for the review!

File: server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java [hunk 2]
```diff
@@ -323,6 +328,9 @@ class NodeTransportHandler implements TransportRequestHandler<NodeTaskRequest> {
...
+            }
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:333]:
```suggestion
```

File: server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java]: ok, removed

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I still think this deserves some kind of test to verify that this is sufficient to make these actions cancellable.

**Comment by benwtrent** [comment]
@DaveCTurner I will see if there is any testing framework that fits the bill. 

I have verified locally that it does (as long as the subsequent calls are cancellable or check for cancellation).

File: server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java [hunk 3]
```diff
@@ -620,6 +628,81 @@ private String getAllTaskDescriptions() {
...
+        CountDownLatch checkLatch = new CountDownLatch(1);
+        CountDownLatch taskLatch = new CountDownLatch(1);
+        ActionFuture<NodesResponse> future = startBlockingTestNodesAction(checkLatch);
+        int numNodes = 2;
+
+        CountDownLatch taskExecutesLatch = new CountDownLatch(numNodes);
+        TestTasksAction[] tasksActions = new TestTasksAction[numNodes];
+        for (int j = 0; j < numNodes; j++) {
+            final int nodeId = j;
+            // Simulate task action that fails on one of the tasks on one of the nodes
```
Comment (benwtrent) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java:643]:
```suggestion
```

**Comment by benwtrent** [comment]
@DaveCTurner added a test that shows how the original action being cancelled propagates down to the node actions

**Review by DaveCTurner — APPROVED** [review]
Nice, LGTM

**Pull request merged**